Packages/com.unity.inputsystem/InputSystem/Controls/QuaternionControl.cs

```diff
@@ -66,17 +67,61 @@ protected override void FinishSetup()
         /// <inheritdoc/>
         public override unsafe Quaternion ReadUnprocessedValueFromState(void* statePtr)
         {
-            return new Quaternion(x.ReadValueFromState(statePtr), y.ReadValueFromState(statePtr), z.ReadValueFromState(statePtr),
-                w.ReadUnprocessedValueFromState(statePtr));
+            EnsureOptimizationTypeHasNotChanged();
+
+            switch (m_OptimizedControlDataType)
+            {
+                case InputStateBlock.kFormatQuaternion:
+                    return *(Quaternion*)((byte*)statePtr + (int)m_StateBlock.byteOffset);
+                default:
+                    return new Quaternion(
+                        x.ReadValueFromState(statePtr),
+                        y.ReadValueFromState(statePtr),
+                        z.ReadValueFromState(statePtr),
+                        w.ReadUnprocessedValueFromState(statePtr));
+            }
         }
 
         /// <inheritdoc/>
         public override unsafe void WriteValueIntoState(Quaternion value, void* statePtr)
         {
-            x.WriteValueIntoState(value.x, statePtr);
-            y.WriteValueIntoState(value.y, statePtr);
-            z.WriteValueIntoState(value.z, statePtr);
-            w.WriteValueIntoState(value.w, statePtr);
+            EnsureOptimizationTypeHasNotChanged();
+
+            switch (m_OptimizedControlDataType)
+            {
+                case InputStateBlock.kFormatQuaternion:
+                    *(Quaternion*)((byte*)statePtr + (int)m_StateBlock.byteOffset) = value;
+                    break;
+                default:
+                    x.WriteValueIntoState(value.x, statePtr);
+                    y.WriteValueIntoState(value.y, statePtr);
+                    z.WriteValueIntoState(value.z, statePtr);
+                    w.WriteValueIntoState(value.w, statePtr);
+                    break;
+            }
+        }
+
+        protected override FourCC CalculateOptimizedControlDataType()
+        {
+            if (
+                m_StateBlock.sizeInBits == 128 &&
```

Comment:
Why is this 128, get bit of magic number associations, would it be more clear with something like?
m_StateBlock.sizeInBits == sizeof(float) * 4 + sizeof(overhead....)
or similar?

Reply:
Can fix, but dunno, most game devs know that vec4 is 128 bit, kinda a muscle memory I feel

Reply:
Fixed